File: algorithms/spoj-challenges/PA05_POT/main.cpp
#include <iostream>

long long mod_exp(long long base, long long exponent, long long modulus) {
  long long test;

  for (test = 1; exponent; exponent >>= 1) {
    if (exponent & 1) {
        test = (test * base) % modulus;
    }

    base = (base * base) % modulus;
  }

  return test;
}

int main() {
  std::ios_base::sync_with_stdio(0);

  int n = 0;

  std::cin >> n;

  for (int i = 0; i < n; ++i) {
    long long base;
    long long exponent;

    std::cin >> base >> exponent;
    std::cout << mod_exp(base, exponent, 10) % 10 << std::endl;
  }

  return 0;
}

File: algorithms/spoj-challenges/PA05_POT/main.erl
-module(main).
-export([ main/0 ]).

read_lines(0, Lines) -> Lines;
read_lines(N, Lines) ->
    NewLines =  Lines ++ [ read_line() ],
    read_lines(N - 1, NewLines).

read_line() ->
    { ok, [ A, B ] } = io:fread("", "~d ~d"),
    { A, B }.

mod_exp_calculation(Result, Base, Exponent, Modulus) when Exponent > 0 ->
    NewResult = case Exponent rem 2 of
        1 -> (Result * Base) rem Modulus;
        _ -> Result
    end,

    NewExponent = Exponent div 2,
    NewBase = (Base * Base) rem Modulus,

    mod_exp_calculation(NewResult, NewBase, NewExponent, Modulus);
mod_exp_calculation(Result, _, _, _) -> Result.

mod_exp(Base, Exponent, Modulus) ->
    mod_exp_calculation(1, Base, Exponent, Modulus).

main() ->
    { ok, [ N ] } = io:fread("", "~d"),

    List = read_lines(N, []),
    lists:foreach(fun({ A, B }) -> io:fwrite("~p~n", [ mod_exp(A, B, 10) rem 10 ]) end, List).